### PR TITLE
Fallback to "show usages" from "Goto definition" if the symbol represents a definition itself

### PR DESCRIPTION
**Pull request by tanishiking**
This PR adds a feature that fallbacks to "show usages" from "goto definition" if the symbol at the given text document position represents a definition itself, which is proposed in https://github.com/scalameta/metals/issues/755.

Now, Metals returns the reference locations instead of definition location for `textDocument/definition` if the symbol at the given text document position represents the definition itself.

## VSCode example
In this recording, I used only "go to definition".

Peek 2019-06-17 18-41

<details>
<summary>Here's the pieces of JSON RPC logs while recording</summary>

### When `textDocument/definition` returns `definitonOrReferences`'s reference locations'

```
{
  "jsonrpc": "2.0",
  "id": 51,
  "method": "textDocument/definition",
  "params": {
    "textDocument": {
      "uri": "file:///home/tanishiking/dev/src/github.com/tanishiking/metals/metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala"
    },
    "position": {
      "line": 1430,
      "character": 6
    }
  }
}
{
  "jsonrpc": "2.0",
  "id": 51,
  "result": [
    {
      "uri": "file:///home/tanishiking/dev/src/github.com/tanishiking/metals/metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala",
      "range": {
        "start": {
          "line": 742,
          "character": 6
        },
        "end": {
          "line": 742,
          "character": 28
        }
      }
    },
    {
      "uri": "file:///home/tanishiking/dev/src/github.com/tanishiking/metals/tests/unit/src/main/scala/tests/TestingServer.scala",
      "range": {
        "start": {
          "line": 582,
          "character": 30
        },
        "end": {
          "line": 582,
          "character": 52
        }
      }
    }
  ]
}
```

### When `textDocument/definition` returns the location of its definition

```
{
  "jsonrpc": "2.0",
  "id": 49,
  "method": "textDocument/definition",
  "params": {
    "textDocument": {
      "uri": "file:///home/tanishiking/dev/src/github.com/tanishiking/metals/tests/unit/src/main/scala/tests/TestingServer.scala"
    },
    "position": {
      "line": 582,
      "character": 33
    }
  }
}
{
  "jsonrpc": "2.0",
  "id": 49,
  "result": [
    {
      "uri": "file:///home/tanishiking/dev/src/github.com/tanishiking/metals/metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala",
      "range": {
        "start": {
          "line": 1430,
          "character": 6
        },
        "end": {
          "line": 1430,
          "character": 28
        }
      }
    }
  ]
}
```

</details>

## nvim + coc.nvim example

Peek 2019-06-19 01-50

### CODE REVIEW

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala [hunk 1]
```diff
@@ -820,21 +820,7 @@ class MetalsLanguageServer(
...
-          scribe.info(
-            s"time: found ${result.locations.length} references to symbol '${result.symbol}' in $timer"
-          )
-        }
-      }
-      if (result.symbol.nonEmpty) {
-        compileAndLookForNewReferences(params, result)
-      }
-      result.locations.asJava
+      referencesResult(params).locations.asJava
```
Comment (tanishiking) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala]: Just moved the implementation into `referencesResult` that returns `ReferencesResult` instead of `CompletableFuture[util.List[Location]]`.

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala [hunk 2]
```diff
@@ -1420,6 +1421,58 @@ class MetalsLanguageServer(
...
+        positionOccurrence = definitionProvider.positionOccurrence(
+          source,
+          position,
+          doc
+        )
+        occ <- positionOccurrence.occurrence
+      } yield occ) match {
+        case Some(occ) =>
+          if (occ.role.isDefinition) {
+            val referecneContext = new ReferenceContext(false)
```
Comment (gabro) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala]: typo
```suggestion
            val referenceContext = new ReferenceContext(false)
```

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala [hunk 3]
```diff
@@ -1420,6 +1421,58 @@ class MetalsLanguageServer(
...
+            val refParams = new ReferenceParams(referecneContext)
```
Comment (gabro) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala]:
```suggestion
            val refParams = new ReferenceParams(referenceContext)
```

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala [hunk 4]
```diff
@@ -1420,6 +1421,58 @@ class MetalsLanguageServer(
...
+            refParams.setTextDocument(position.getTextDocument())
+            refParams.setPosition(position.getPosition())
+            val result = referencesResult(refParams)
+            Future.successful(
+              DefinitionResult(
+                locations = result.locations.asJava,
+                symbol = result.symbol,
+                definition = None,
+                semanticdb = None
+              )
```
Comment (tanishiking) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala]: Returns `symbol` in addition to the `locations` because `TestingServer` requires symbol information.
Wrapping those by `DefinitionResult` for structuring the return value, but what I actually want is something like [`Definitionresult | ReferencesResult`](https://dotty.epfl.ch/docs/reference/new-types/union-types.html) haha.

File: tests/unit/src/test/scala/tests/DefinitionSlowSuite.scala [hunk 5]
```diff
@@ -84,9 +84,9 @@ object DefinitionSlowSuite extends BaseSlowSuite("definition") {
            |import java.util.concurrent.Future/*Future.java*/ // unused
            |import scala.util.Failure/*Try.scala*/ // unused
            |import org.scalatest.FunSuite/*FunSuite.scala*/
-           |object MainSuite/*L4*/ extends FunSuite/*FunSuite.scala*/ {
+           |object MainSuite/*<no symbol>*/ extends FunSuite/*FunSuite.scala*/ {
```
Comment (gabro) [inline tests/unit/src/test/scala/tests/DefinitionSlowSuite.scala]: This change means that we were previously returning the symbol itself, whereas now we don't return any symbol. I'm wondering whether we should always fall back to the symbol itself if there are no references.

I'm not feeling strongly about this, just wanted to discuss it, since it's a minor change from the previous behavior.

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala [hunk 6]
```diff
@@ -1420,6 +1421,58 @@ class MetalsLanguageServer(
...
+            )
+          } else {
+            definitionResult(position, token)
+          }
+        case None =>
+          warnings.noSemanticdb(source)
+          // Even if it failed to retrieve the symbol occurrence from semanticdb,
+          // try to find its definitions from presentation compiler.
+          definitionResult(position, token)
```
Comment (tanishiking) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala]: Call `definitionResult` even if it failed to retrieve the symbol information from semanticdb because `DefinitionProvider#definition` [fallback to finding definitions from a presentation compiler](https://github.com/scalameta/metals/blob/50912543e057d958e4a336a6028f74ebca71271e/metals/src/main/scala/scala/meta/internal/metals/DefinitionProvider.scala#L62) if no definition locations found from semanticdb.

File: tests/unit/src/test/scala/tests/DefinitionSlowSuite.scala (hunk 5 above)
Comment (tanishiking) [inline tests/unit/src/test/scala/tests/DefinitionSlowSuite.scala]: Now, the symbol that represents a definition shows the locations of its references, instead of their locations themselves.
In this case, `MainSuite` shows `<no symbol>` because `MainSuite` represents a definition, and no one refers to `MainSuite`.

File: tests/unit/src/test/scala/tests/DefinitionSlowSuite.scala [hunk 7]
```diff
@@ -123,9 +123,9 @@ object DefinitionSlowSuite extends BaseSlowSuite("definition") {
...
            |import scala.util.Failure/*Try.scala*/ // unused
            |import org.scalatest.FunSuite/*FunSuite.scala*/
-           |object MainSuite/*L6*/ extends FunSuite/*FunSuite.scala*/ {
+           |object MainSuite/*<no symbol>*/ extends FunSuite/*FunSuite.scala*/ {
            |  test/*FunSuiteLike.scala*/(testName/*<no symbol>*/) {
-           |    val condition/*L8*/ = Main/*Main.scala:5*/.message/*<no symbol>*/.contains/*String.java*/("Hello")
+           |    val condition/*L9*/ = Main/*Main.scala:5*/.message/*<no symbol>*/.contains/*String.java*/("Hello")
            |    assert/*Assertions.scala*/(condition/*L8*/)
            |  }
            |}
```
Comment (tanishiking) [inline tests/unit/src/test/scala/tests/DefinitionSlowSuite.scala]: I tweaked this test case for fixing some that are related to the change, but this test case is still failing (even in `master` branch)...

```
=======
=> Diff
=======
---∙
+++∙
@@ -2,3 +2,3 @@
 package a
-import java.util.concurrent.Future/*<no symbol>*/ // unused
+import java.util.concurrent.Future/*Future.java*/ // unused
 import scala.util.Failure/*Try.scala*/ // unused
@@ -6,3 +6,3 @@
   val message/*L4*/ = Message/*Message.java:1*/.message/*Message.java:2*/
-  new java.io.PrintStream/*<no symbol>*/(new java.io.ByteArrayOutputStream/*<no symbol>*/())
+  new java.io.PrintStream/*PrintStream.java*/(new java.io.ByteArrayOutputStream/*ByteArrayOutputStream.java*/())
   println/*Predef.scala*/(message/*L4*/)
@@ -11,3 +11,3 @@
 package a
-import java.util.concurrent.Future/*<no symbol>*/ // unused
+import java.util.concurrent.Future/*Future.java*/ // unused
 import scala.util.Failure/*Try.scala*/ // unused
@@ -16,3 +16,3 @@
   test/*FunSuiteLike.scala*/("a") {
-    val condition/*L6*/ = Main/*Main.scala:3*/.message/*Main.scala:4*/.contains/*<no symbol>*/("Hello")
+    val condition/*L6*/ = Main/*Main.scala:3*/.message/*Main.scala:4*/.contains/*String.java*/("Hello")
     assert/*Assertions.scala*/(condition/*L6*/)
  tests.DiffAssertions$.colored(DiffAssertions.scala:103)
  tests.BaseSuite.assertNoDiff(BaseSuite.scala:102)
  tests.DefinitionSlowSuite$.$anonfun$new$5(DefinitionSlowSuite.scala:71)
  scala.util.Success.$anonfun$map$1(Try.scala:255)
  scala.util.Success.map(Try.scala:213)
  scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
  scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
  scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
  scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
  java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
  java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
  java.lang.Thread.run(Thread.java:748)
```

I'm not sure why we cannot resolve the location of java resources :thinking:

**Comment by tanishiking** [comment]
Thank you for looking at this @gabro !

> In #755 I was worried about how clients dealt with N results from textDocument/definition, have you perhaps tried this with clients other than VSCode?

Not yet..., and I'm going to try it on other editors :)
I think I can test it easily on `vim + coc.nvim`, but I've never setup other text editors than vim and VSCode, so it may take some time.

File: tests/unit/src/test/scala/tests/DefinitionSlowSuite.scala (hunk 5 above)
Comment (olafurpg) [reply tests/unit/src/test/scala/tests/DefinitionSlowSuite.scala]: I am inclined to fall back to the original behaviour in this case and return the definition result. The reason is to avoid confusing messages like "No definition found for MainSuite"

Screenshot 2019-06-18 at 15 00 59

File: tests/unit/src/test/scala/tests/DefinitionSlowSuite.scala (hunk 7 above)
Comment (tanishiking) [reply tests/unit/src/test/scala/tests/DefinitionSlowSuite.scala]: Ah, this was just the problem on my local computer, `sudo apt install openjdk-8-source` fixed this :)

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala (hunk 4 above)
Comment (tanishiking) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala]: That's right! But this time, let me leave the return value's type `Future[DefinitionResult]` because nesting `Either` inside `Future` makes things complicated and feels like to use something like `EitherT` ...

File: tests/unit/src/test/scala/tests/DefinitionSlowSuite.scala [hunk 8]
```diff
@@ -134,6 +134,63 @@ object DefinitionSlowSuite extends BaseSlowSuite("definition") {
...
+      )
+      _ <- server.didOpen("a/src/main/scala/a/A.scala")
+      _ <- server.didOpen("b/src/main/scala/a/B.scala")
+      _ = assertNoDiff(client.workspaceDiagnostics, "")
+      _ = assertNoDiff(
+        server.workspaceDefinitions,
+        """|/a/src/main/scala/a/A.scala
+           |package a
+           |object A/*B.scala:3*/ {
+           |  val name/*;B.scala:3;L4*/ = "John"
```
[the remark this answers is not in the document]
Comment (tanishiking) [reply tests/unit/src/test/scala/tests/DefinitionSlowSuite.scala]: sorted https://github.com/scalameta/metals/pull/775/files#diff-7b39581a6c581b96e5939753620ed527R604

File: tests/unit/src/test/scala/tests/DefinitionSlowSuite.scala (hunk 5 above)
Comment (tanishiking) [reply tests/unit/src/test/scala/tests/DefinitionSlowSuite.scala]: Sounds reasonable! done https://github.com/scalameta/metals/pull/775/files#diff-f8c05eebbf12c9c21a7d568f09b500eaR1457

File: tests/unit/src/test/scala/tests/DefinitionSlowSuite.scala [hunk 9]
```diff
@@ -217,6 +274,9 @@ object DefinitionSlowSuite extends BaseSlowSuite("definition") {
            |.metals/readonly/cats/Contravariant.scala:10:45: information: <error> takes no type parameters, expected: one
            |  def compose[G[_]: Contravariant]: Functor[λ[α => F[G[α]]]] =
            |                                            ^
+           |.metals/readonly/cats/Contravariant.scala:13:15: information: not found: value Contravariant
+           |      val G = Contravariant[G]
+           |              ^^^^^^^^^^^^^
```
Comment (tanishiking) [inline tests/unit/src/test/scala/tests/DefinitionSlowSuite.scala]: Hmm, I'm not sure why this change effects like this ...

**Comment by tanishiking** [comment]
Now, I fixed things based on the review and tests passed :)

- I left a comment on the test for `annotation`
  - https://github.com/scalameta/metals/pull/775/files#r294970446
- Test manually on `nvim + coc.nvim`
  - I'm going to test on other editors in this week https://github.com/scalameta/metals/pull/775#issuecomment-503053512

File: tests/unit/src/test/scala/tests/DefinitionSlowSuite.scala (hunk 7 above)
Comment (olafurpg) [reply tests/unit/src/test/scala/tests/DefinitionSlowSuite.scala]: Our docs should probably mention this for ubuntu users 🤔

File: tests/unit/src/test/scala/tests/DefinitionSlowSuite.scala (hunk 9 above)
Comment (olafurpg) [reply tests/unit/src/test/scala/tests/DefinitionSlowSuite.scala]: Hmm, I'm not sure either. The only thing I can think of is that we're not calling this line here in some cases https://github.com/scalameta/metals/blob/4cfc3a51daeb457a8f1e26bea4196104de517380/metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala#L1501

Either way, the change looks benign so let's keep the diff.
